packages/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -273,6 +282,15 @@ export const GridRootStyles = styled('div', {
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
       outline: `solid ${t.palette.primary.main} 1px`,
     },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="1"]`]: {
+      borderTopLeftRadius: 'var(--unstable_DataGrid-radius)',
+    },
+    [`&.${c['root--noToolbar']} [aria-rowindex="1"] [aria-colindex="${visibleColumns.length}"]`]: {
+      borderTopRightRadius:
+        !dimensions.hasScrollY && dimensions.scrollbarSize > 0
+          ? 'var(--unstable_DataGrid-radius)'
+          : undefined,
```

Comment:
```suggestion
        !dimensions.hasScrollY || dimensions.scrollbarSize === 0
          ? 'var(--unstable_DataGrid-radius)'
          : undefined,
```

I believe this is the correct condition.
With the suggestion above applied, seems to work fine:


https://github.com/user-attachments/assets/a4c7dbcb-7c44-4daa-b724-6bb7392f1a14